src/provider/shplonk.rs

```diff
@@ -284,23 +266,35 @@ where
     // P_i+1(r^2) == (1 - point_i) * P_i_even + point_i * P_i_odd -> should hold, according to Gemini transformation
     let mut point = point.to_vec();
     point.reverse();
+
     #[allow(clippy::disallowed_methods)]
-    for (index, ((eval_r, eval_minus_r), eval_r_squared)) in evals_at_r
-      .iter()
-      .zip_eq(evals_at_minus_r.iter())
-      // TODO: Ask Adrian if we need evals_at_r_squared[0] for some additional checks
-      .zip(evals_at_r_squared[1..].iter())
+    let verification_failed = pi.evals[0]
+      .par_iter()
+      .chain(&[*P_of_x])
+      .zip_eq(pi.evals[1].par_iter().chain(&[*P_of_x]))
+      .zip(pi.evals[2][1..].par_iter().chain(&[*P_of_x]))
       .enumerate()
-    {
-      let even = (*eval_r + eval_minus_r) * (E::Fr::from(2).invert().unwrap());
-      let odd = (*eval_r - eval_minus_r) * ((E::Fr::from(2) * r).invert().unwrap());
+      .any(|(index, ((eval_r, eval_minus_r), eval_r_squared))| {
+        let even = (*eval_r + eval_minus_r) * (E::Fr::from(2).invert().unwrap());
+        let odd = (*eval_r - eval_minus_r) * ((E::Fr::from(2) * r).invert().unwrap());
 
-      if *eval_r_squared != ((E::Fr::ONE - point[index]) * even) + (point[index] * odd) {
-        return Err(NovaError::ProofVerifyError);
-      }
+        *eval_r_squared != ((E::Fr::ONE - point[index]) * even) + (point[index] * odd)
+      });
+
+    if verification_failed {
+      return Err(NovaError::ProofVerifyError);
     }
```

Comment:
I think these checks would be clearer if we iterated over `i`, to show the dependency between 
`evals_r[i], evals_negative_r[i], evals_r_squared[i+1]`

Reply:
Field inversions are also very expensive. An easy improvement here would be to compute `1/2` and `1/2r` once before the start of the loop. 

The next best solution would be to re-write the equality check by clearing all denominators
$$
2 r P_{i+1}(r^2) = r(1-x_i)(P_i(r) + P_i(-r)) + x_i(P_i(r) - P_i(-r))
$$

(please double the math 🙏)

Reply:
I think comment above is self-explanatory - but happy to rewrite if you insist